cooperative_lanechange/src/cooperative_lanechange.cpp

```diff
@@ -0,0 +1,867 @@
+        double relative_velocity = current_speed_ - veh2_speed;
+        double desired_gap = desired_time_gap_ * relative_velocity;      
+         
+        if(current_gap < desired_gap){
+            return true;
+        }
+        //plan lanechange without filling in response
+        std::vector<cav_msgs::TrajectoryPlanPoint> planned_trajectory_points = plan_lanechange(req);
+
+        //send mobility request
+        //Planning for first lane change maneuver
+        cav_msgs::MobilityRequest request = create_mobility_request(planned_trajectory_points, maneuver_plan[0]);
+        outgoing_mobility_request_.publish(request);
+        ros::Time request_sent_time = ros::Time::now();
+        cav_msgs::LaneChangeStatus lc_status_msg;
+        lc_status_msg.status = cav_msgs::LaneChangeStatus::PLAN_SENT;
+        lc_status_msg.description = "Requested lane merge";
+        lanechange_status_pub_.publish(lc_status_msg);
+
+        //if ack mobility response, send lanechange response
+        if(is_lanechange_accepted_){
+            cav_msgs::TrajectoryPlan trajectory_plan;
+            trajectory_plan.header.frame_id = "map";
+            trajectory_plan.header.stamp = ros::Time::now();
+            trajectory_plan.trajectory_id = boost::uuids::to_string(boost::uuids::random_generator()());
+
+            trajectory_plan.trajectory_points = planned_trajectory_points;
+            trajectory_plan.initial_longitudinal_velocity = std::max(req.vehicle_state.longitudinal_vel, minimum_speed_);
+            resp.trajectory_plan = trajectory_plan;
+
+            for (int i=0; i<req.maneuver_plan.maneuvers.size(); i++){
+                if (req.maneuver_plan.maneuvers[i].type == cav_msgs::Maneuver::LANE_CHANGE){
+                    resp.related_maneuvers.push_back(i);
+                    break;
+                }
+            }
+            resp.maneuver_status.push_back(cav_srvs::PlanTrajectory::Response::MANEUVER_IN_PROGRESS);
+
+        }
+        else{
+            ros::Time planning_end_time = ros::Time::now();
+            ros::Duration passed_time = planning_end_time - request_sent_time;
+            if(passed_time.toSec() >= lanechange_time_out_){
+                cav_msgs::LaneChangeStatus lc_status_msg;
+                lc_status_msg.status = cav_msgs::LaneChangeStatus::TIMED_OUT;
+                lc_status_msg.description = "Request timed out for lane merge";
+                lanechange_status_pub_.publish(lc_status_msg);
+            }
+ 
+        }
+
+        return true;
+    }
+    
+    cav_msgs::MobilityRequest CooperativeLaneChangePlugin::create_mobility_request(std::vector<cav_msgs::TrajectoryPlanPoint>& trajectory_plan, cav_msgs::Maneuver maneuver){
+
+        cav_msgs::MobilityRequest request_msg;
+        cav_msgs::MobilityHeader header;
+        header.sender_id = sender_id_;
+        header.recipient_id = DEFAULT_STRING_;      //To do- find way to query recipient id
```

Comment:
We don't necessarily need to implement it this way, but in CARMA 2 this mobility request was treated as a broadcast most of the time, not a uni-cast. We used it mostly to say "Heres the path I would like to do, but is currently not possible, can anyone in the surrounding area do something (read: slow down, etc.) to allow me to complete this path. Any vehicle that didn't feel it was relevant would simply not respond, any vehicle that deemed it to be a fundamentally unsafe action would send a MobilityResponse NACK, and any vehicle that was willing and able to help would send a MobilityResponse ACK. If we got at least 1 ACK, and non NACKs we would proceed with the plan, otherwise we'd abort that operation or try a different proposed plan.

Just a thought.

If you think it makes sense to treat this as a uni-cast negotiation, then we'll need to query the world model of the vehicle we identified as being in conflict with us and get it's BSM data that way.